cache/edge_lambdas/src/rejection.ts

```diff
@@ -0,0 +1,205 @@
+ *      with overly expensive spam checks.
+ *
+ * == How these heuristics were developed ==
+ *
+ * I downloaded the CloudFront logs for a 2-week-long period, and looked at any request
+ * that had a `query` parameter (which is where a lot of the spam comes in).
+ *
+ * Then I wrote a script that would try a candidate heuristic over these requests,
+ * and report:
+ *
+ *    - the total number of requests that were accepted/rejected
+ *    - the average time taken to process each request
+ *    - a random sample of accepted/rejected requests
+ *
+ * I ran this multiple times to refine the heuristic, until a good number of the sample
+ * of accepted requests looked like bona-fide traffic.
+ *
+ */
+
+import {
+  CloudFrontRequest,
+  CloudFrontRequestEvent,
+  CloudFrontResponse,
+} from 'aws-lambda';
+
+// Returns true if this request looks like a bot.
+//
+// A lot of spam requests will spoof their User-Agent header to look like a popular
+// web crawler, e.g. Googlebot or Bingbot.  Because a person would never spoof their
+// User-Agent this way, we can use this as an excuse to apply more stringent checks.
+const isBotRequest = (request: CloudFrontRequest): boolean => {
+  const userAgent = request.headers['user-agent']?.[0]?.value || '';
+
+  return (
+    userAgent.toLowerCase().includes('bot') ||
+    userAgent.toLowerCase().includes('search.yahoo')
+  );
+};
+
+// 253,376 requests rejected
+// 6,688,393 requests accepted
+
+export const looksLikeSpam = (event: CloudFrontRequestEvent): boolean => {
+  const cf = event.Records[0].cf;
+  const request = cf.request;
+
+  const requestParams = new URLSearchParams(request.querystring);
+  const query = requestParams.getAll('query').join(' ');
+
+  // Look for the `query` URL parameter.
+  //
+  // The majority of our spam comes from people plugging in values to the
+  // search form and stuffing bogus values into the query parameter.
+  // If there isn't a query parameter on this request, it's probably not spam.
+  if (query === '') {
+    return false;
+  }
+
+  // Look for spammy keywords and emoji.
+  //
+  // We can't solely rely on the presence of these words/characters to indicate spam,
+  // but they're a clue -- usually accompanied by long strings of Chinese characters
+  // and a URL to a sketchy-looking site.
+  //
+  // To avoid penalising real users, we only treat this text as spam if the user
+  // has self-identified as a bot/crawler in their User-Agent header.
+  //
+  // In my analysis, this flagged ~41% of all search traffic.
+  const spamKeywords = [
+    'cash',
+    'casino',
+    'chatgpt',
+    'crypto',
+    'ddos',
+    'free',
+    'fun',
+    'game',
+    'give away',
+    'give-away',
+    'giveaway',
+    'play',
+    'poker',
+    'telegram',
+    'whatsapp',
+    'win',
+    '➡️',
+    '☀️',
+    '⏩',
+    '⛔',
+    '⚽',
+    '✅',
+    '🐯',
+    '🍀',
+    '㊙',
+    '✔️',
+    '⭐',
+    '👈',
+    '👉',
+  ];
```

Comment:
Hmmmm. I'm thinking about false positives here, as I wouldn't want a real search crawler to be rejected for a canned search for (eg) `play`. This is less significant now we have concepts pages, but I wonder if this list is a bit aggressive? As with my other comment, also wondering about the overlap between this 41% and the 70% below...

Reply:
Good point, especially as we literally used to have an exhibition with Play in the title.

It's catching ~6.5% of requests that aren't caught by the unusual character filter.

How would you feel about making the list less aggressive? Remove all of the following:

    play free fun win giveaway give-away cash game

especially as there's no stemming – e.g. 'window' would also match – and leave the stuff which is less ambiguously spammy:

    crypto casino chatgpt whatsapp ddos poker
    # plus all the emoji

That gets us most of the benefits (~5.7% extra requests) without being quite as aggressive.

Reply:
Yep, happy with that